### PR TITLE
Feature/us0000 do not refresh token unless expired

### PR DESCRIPTION
**Pull request by mjvcallibrity**

### CODE REVIEW

File: Gateway/crds-angular/Services/AuthTokenExpiryService.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  public class AuthTokenExpiryService : IAuthTokenExpiryService
+  {
+    public bool IsAuthtokenCloseToExpiry(HttpRequestHeaders headers)
+    {
+      const int expirationBufferInSeconds = 60;
+
+      IEnumerable<string> authorizationHeaders;
+      headers.TryGetValues("Authorization", out authorizationHeaders);
+
+      if (authorizationHeaders == null || !authorizationHeaders.Any())
```
Comment (mjvcallibrity) [inline Gateway/crds-angular/Services/AuthTokenExpiryService.cs]: This logic check could be pulled out into a private function and given a name that is more descriptive of what exactly this check means. Something like `areAuthorizationHeadersNotPresent(authorizationHeaders)`

File: crossroads.net/core/services/session_service.js [hunk 2]
```diff
@@ -178,52 +208,56 @@
...
-      if (this.exists('redirectUrl') !== undefined) {
+      if (this.exists("redirectUrl") !== undefined) {
         return true;
       }
       return false;
     };
 
     vm.verifyAuthentication = (event, stateName, stateData, stateToParams) => {
       if (vm.isActive()) {
+        debugger;
```
[the remark this answers is not in the document]
Comment (Doug-Shannon) [reply crossroads.net/core/services/session_service.js]: left it in just to see if you would find it --- you passed, well done ;)

File: Gateway/crds-angular.test/Services/GoogleMapsDistanceMatrixAddressProximityServiceTest.cs [hunk 3]
```diff
@@ -75,7 +76,9 @@ public void TestGetProximity()
             Assert.IsNotNull(result);
             Assert.AreEqual(1, result.Count);
             Assert.IsNotNull(result[0]);
-            Assert.AreEqual(16.6, (double)result[0], .009);
+            double resultProximity = (double)result[0];
+            double acceptableErrrorDistance = .5d;
+            Assert.AreEqual(16.6, resultProximity, acceptableErrrorDistance);
```
[the remark this answers is not in the document]
Comment (vchtelmakh) [reply Gateway/crds-angular.test/Services/GoogleMapsDistanceMatrixAddressProximityServiceTest.cs]: This test was always broken. I fixed it. I talked to Phil to ensure this was an acceptable rounding error.

File: crossroads.net/core/services/session_service.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mjvcallibrity) [reply crossroads.net/core/services/session_service.js]: Done!

File: crossroads.net/core/services/session_service.js [hunk 4]
```diff
@@ -178,52 +208,56 @@
...
-          $cookies.put('username', user.username);
-          vm.enableReactiveSso(event, stateName, stateData, stateToParams);
-          vm.restoreImpersonation();
-        }, (response) => {
-          if (response.status !== -1) {
-            vm.clearAndRedirect(event, stateName, stateToParams);
+        }).then(
+          response => {
+            var user = response.data;
+            $rootScope.userid = user.userId;
```
Comment (vchtelmakh) [inline crossroads.net/core/services/session_service.js]: Can we make this `$rootScope.userid` camelcase or is it set in too many places?

File: crossroads.net/spec-core/services/session.service.spec.js [hunk 5]
```diff
@@ -76,8 +76,15 @@ describe('Session Service', function () {
     });
 
     it('should set credentials when login is detected', function () {
+      headersResp = {}
+      headersResp[cookieNames.SESSION_ID] = 'abcdef';
       $cookies.put(cookieNames.SESSION_ID, mockResponse.userToken);
-      Backend.expectGET(window.__env__['CRDS_GATEWAY_CLIENT_ENDPOINT'] + 'api/authenticated').respond(200, mockResponse);
+      // Backend.expectGET(window.__env__['CRDS_GATEWAY_CLIENT_ENDPOINT'] + 'api/authenticated').respond(200, mockResponse);
```
Comment (vchtelmakh) [inline crossroads.net/spec-core/services/session.service.spec.js]: We should delete the comment.

File: crossroads.net/core/services/session_service.js [hunk 6]
```diff
@@ -135,29 +166,28 @@
...
       $cookies.remove(cookieNames.REFRESH_TOKEN);
       $cookies.remove(cookieNames.IMPERSONATION_ID);
-      $cookies.remove('userId');
-      $cookies.remove('username');
-      $cookies.remove('family');
-      $cookies.remove('age');
+      $cookies.remove("userId");
+      $cookies.remove("username");
+      $cookies.remove("family");
+      $cookies.remove("age");
```
Comment (vchtelmakh) [inline crossroads.net/core/services/session_service.js]: Another nitpicky "not necessary to change" thing, but why the double quotes?

File: crossroads.net/core/components/interceptor.js [hunk 7]
```diff
@@ -1,38 +1,50 @@
...
 
-      response: function(response) {
-        if (response.headers('refreshToken')) {
-          var Session = $injector.get('Session');
-          Session.refresh(response);
+      response: function (response) {
+        const url = response.config.url;
+        const isHttp = url.startsWith("http://") || url.startsWith("https://");
+        const expDate = new Date();
+        const sessionLength = 1800000;
```
Comment (vchtelmakh) [inline crossroads.net/core/components/interceptor.js]: Would be nice to have units on the `sessionLength` - had to think about it for a second to realize that its `sessionLengthMs`

File: crossroads.net/core/services/session_service.js (hunk 4 above)
Comment (mjvcallibrity) [reply crossroads.net/core/services/session_service.js]: It currently is camelCase.

File: crossroads.net/spec-core/services/session.service.spec.js (hunk 5 above)
Comment (mjvcallibrity) [reply crossroads.net/spec-core/services/session.service.spec.js]: I'm on it. Thanks for catching that.

File: crossroads.net/core/components/interceptor.js (hunk 7 above)
Comment (mjvcallibrity) [reply crossroads.net/core/components/interceptor.js]: 👍 Can do.

File: crossroads.net/core/services/session_service.js (hunk 4 above)
Comment (mjvcallibrity) [reply crossroads.net/core/services/session_service.js]: Whoops. Sorry. Now I see what you mean. `userid` should be `userId`.

File: Gateway/crds-angular.test/Services/GoogleMapsDistanceMatrixAddressProximityServiceTest.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mjvcallibrity) [reply Gateway/crds-angular.test/Services/GoogleMapsDistanceMatrixAddressProximityServiceTest.cs]: 👍 Thanks!

File: crossroads.net/core/services/session_service.js (hunk 4 above)
Comment (vchtelmakh) [reply crossroads.net/core/services/session_service.js]: ` $rootScope.userid` -> `$rootScope.userId` Not a big deal though.

File: crossroads.net/core/services/session_service.js (hunk 6 above)
Comment (mjvcallibrity) [reply crossroads.net/core/services/session_service.js]: I'll have to check with Doug, but I think you are right. Technically those _should_ be single quotes.

**Pull request merged**